processor/src/main/java/org/mapstruct/ap/internal/util/AbstractElementUtilsDecorator.java

```diff
@@ -34,20 +36,91 @@
 
 public abstract class AbstractElementUtilsDecorator implements ElementUtils {
 
+    private static final Method MODULE_OF_METHOD;
+    private static final Method GET_TYPE_ELEMENT_WITH_MODULE_METHOD;
+    private static final Method GET_PACKAGE_ELEMENT_WITH_MODULE_METHOD;
+
+    static {
```

Comment:
why not use the java version? >8 this method is available. == 8 not. I'm not in favour of using reflection if that's not strictly needed. So.. you could deduct the java version here (I guess when we've some logic already when we write the @Generated annotation).

You'll need to compile MapStruct to java9. There are some questions arising from this: what would happen if you use the mapstruct processor compiled to 9 in an javac 8 compiler (runtime)?

Reply:
This is not only about the `@Generated` notation. It is also about types that are available or not, they might be available during the annotation processing, but not in the actual module of the mapper (if someone uses module-info.java).

In order for us to be able to use the methods during compile time then we will need to either use a Multi Release Jar where we will offload this or change the baseline of MapStruct.

Reply:
The support for Jakarta in #2738 might also be affected by this. e.g. you might have both javax and jakarta available, but in your module descriptor you are only using jakarta, in this case this new mechanism will properly see only the jakarta packages.

Reply:
>> This is not only about the @Generated notation

I understand that. I was just referring to the code that establishes the runtime java version. 

So in short:
* what would happen if we go to java9 for compiling the processor, especially when use the processor in a javac v8? Would that work if we don't call the java9 methods?
* So in java9 the required methods on `Elements` are available. We only should not call those in a javac 8 environment. 
* we could use the static part to establish if we run the processor in java>8 or java==8. The checks in the methods would than boil down to selecting the proper method.

Reply:
>what would happen if we go to java9 for compiling the processor, especially when use the processor in a javac v8? Would that work if we don't call the java9 methods?

In order to achieve what you are saying we need to change the baseline, which means that the bytecode will have a newer byte code version and you won't be able to run the processor on javac8.

There are 3 options for we supporting Java Modules properly:

* Use reflection like it is done in this PR and like we do for Records
* Use Multi Release Jar and have a different implementation for Java 9+ for this particular class
* Change our baseline to Java 11 or 17. This option is out of the question for now, perhaps in 2.0 we can change our baseline as well. We could still target the generated code to be correct on Java 8, but you would have to use our baseline to run the processor.

Reply:
I agree that we should not change the entire baseline. Java 8 is too much in use right now.